Split up and refactor some NixOS tests

aszlig opened the pull request:
This splits a few NixOS tests (namely Chromium, VirtualBox and the networking tests) into several subtests that are exposed via attributes.

The networking tests were already split up but they didn't expose an attribute set of available tests but used a function attribute to specify the resulting test instead.

A new function `callSubTests` in `nixos/release.nix` is now responsible for gathering subtests, which is also used for the `installer` and `boot` tests. The latter is now placed in a `tests.boot.*` namespace rather than "polluting" the `tests` attribute set with its subtest.

I've added a jobset on my Hydra at:

https://headcounter.org/hydra/jobset/aszlig/nixos-tests-splitup

This is a comparison between `nixpkgs` master (at 11327ac849493d90721f9a36e5f98033dbdeea11) and my changes:

https://headcounter.org/hydra/eval/310933?compare=310931

Cc: @edolstra, @wkennington, @bobvanderlinden

aszlig committed:
nixos/tests/virtualbox: Split up subtests

Now subtests are separate derivations, because the individual tests do
not depend on state from previous test runs.

This has the advantage that it's easier to run individiual tests and
it's also easier to pinpoint individual tests that randomly fail.

I ran all of these tests locally and they still succeed.

Signed-off-by: aszlig <aszlig@redmoonstudios.org>

aszlig committed:
nixos/release.nix: Add a callSubTests function

This should de-clutter the various redundant lines of callTest's on
subtests so that every main test file should have only one line with a
callSubTests instead.

Overrides work the same way as callTest, except that if the system
attribute is explicitly specified we do not generate attributes for all
available systems.

Signed-off-by: aszlig <aszlig@redmoonstudios.org>
Cc: @edolstra

aszlig committed:
nixos/tests/networking: Expose subtests via attrs

So far the networking test expression only generated a single test
depending on the passed "test" attribute. This makes it difficult to
autodiscover the subtests with our shiny new callSubTests function.

This change essentially doesn't change the behaviour of the subtests but
rather exposes them as an attribute set instead of relying on a
particular input argument.

The useNetworkd argument still exists however.

Signed-off-by: aszlig <aszlig@redmoonstudios.org>
Cc: @wkennington

aszlig committed:
nixos/release: Use callSubTests for subtests

Should clean up a lot of these redundant lines for various sub-tests.

Note that the tests.boot* are now called tests.boot.boot*, but otherwise
all the test attribute names should stay the same.

Signed-off-by: aszlig <aszlig@redmoonstudios.org>
Cc: @edolstra
Cc: @wkennington
Cc: @bobvanderlinden

aszlig committed:
nixos/tests/chromium: Split up into subtests

This makes it easier to test just a specific channel rather than to
force testing all builds down the users/testers throat. Especially this
makes it easier to test NixOS channel upgrades only against the Chromium
stable channel instead of just removing the beta/dev channels from the
tests entirely (as done in 69ec09f38aa1f1d37baec73ebdf9cf5f21050f94).

Signed-off-by: aszlig <aszlig@redmoonstudios.org>

domenkozar commented:
Awesome. Chromium is the most essential tool on my machine, I'd also like to have `chromium.stable` as release blockers in 16.03

bobvanderlinden commented:
Looks like that cleans things up quite a bit! Just one aesthetics note, the boot tests could now be renamed from `boot.bootBiosCdrom` to `boot.biosCdrom` in https://github.com/aszlig/nixpkgs/blob/nixos-tests-splitup/nixos/tests/boot.nix#L33. That makes them more consistent with the other tests.

edolstra commented:
:+1:

aszlig committed:
nixos/tests: Remove tests.boot.boot* prefixes

As @bobvanderlinden suggests in #13585:

"Looks like that cleans things up quite a bit! Just one aesthetics note,
the boot tests could now be renamed from boot.bootBiosCdrom to
boot.biosCdrom in nixos/tests/boot.nix:L33.

That makes them more consistent with the other tests."

Signed-off-by: aszlig <aszlig@redmoonstudios.org>
